madsim/src/sim/net/tcp/network.rs

```diff
@@ -0,0 +1,267 @@
+    conn: HashMap<ConnId, Connection>,
+    clogged_node: HashSet<NodeId>,
+    clogged_link: HashSet<(NodeId, NodeId)>,
+    conn_cnt: AtomicU64,
+}
+
+impl TcpNetwork {
+    pub(crate) fn new(rand: GlobalRng, time: TimeHandle, config: Config) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(Inner {
+                rand,
+                time,
+                config,
+                accept: HashMap::default(),
+                conn: HashMap::default(),
+                clogged_node: HashSet::default(),
+                clogged_link: HashSet::default(),
+                conn_cnt: 0.into(),
+            })),
+        }
+    }
+
+    pub(crate) fn update_config(&self, f: impl FnOnce(&mut Config)) {
+        let mut inner = self.inner.lock().unwrap();
+        f(&mut inner.config);
+    }
+
+    pub(crate) fn reset_node(&self, node: NodeId) -> Result<(), String> {
+        trace!("reset node {}", node);
+        let mut inner = self.inner.lock().unwrap();
+        let mut drop_ids = vec![];
+
+        for (conn_id, conn) in &inner.conn {
+            if conn.src == node || conn.dst == node {
+                drop_ids.push(*conn_id);
+            }
+        }
+
+        if drop_ids.is_empty() {
+            Err("there is no node to reset".to_string())
+        } else {
+            for id in drop_ids {
+                inner.conn.remove(&id);
+            }
+
+            Ok(())
+        }
+    }
+
+    pub(crate) fn drop_connection(&self, conn: ConnId) -> io::Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        let removed = inner.conn.remove(&conn).is_some();
+        if removed {
+            Ok(())
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "connection not exist or already shutdown".to_string(),
+            ))
+        }
+    }
+
+    pub(crate) fn clog_node(&self, id: NodeId) {
+        trace!("tcp clog: {id}");
+        self.inner.lock().unwrap().clogged_node.insert(id);
+    }
+
+    pub(crate) fn unclog_node(&self, id: NodeId) {
+        trace!("tcp unclog: {id}");
+        self.inner.lock().unwrap().clogged_node.remove(&id);
+    }
+
+    pub(crate) fn clog_link(&self, src: NodeId, dst: NodeId) {
+        trace!("clog: {src} -> {dst}");
+        self.inner.lock().unwrap().clogged_link.insert((src, dst));
+    }
+
+    pub(crate) fn unclog_link(&self, src: NodeId, dst: NodeId) {
+        trace!("tcp unclog: {src} -> {dst}");
+        self.inner.lock().unwrap().clogged_link.remove(&(src, dst));
+    }
+
+    pub(crate) fn listen(
+        &self,
+        id: NodeId,
+        addr: SocketAddr,
+    ) -> Result<mpsc::Receiver<(ConnId, ConnId)>, String> {
+        let (tx, rx) = mpsc::channel(0);
+        let is_inuse = self
+            .inner
+            .lock()
+            .unwrap()
+            .accept
+            .insert(addr, (id, tx))
+            .is_some();
```

Comment:
If address is in use, the old connection will be replaced. It seems not right?

Reply:
Yes, it's indeed a buggy implement, I'll fix it.